samples/compiler_plugins/xnnpack/test/stablehlo-to-xnnpack.mlir

```diff
@@ -1,29 +1,45 @@
+// CHECK:           %{{.*}} = xnnpack.fully_connected_nc_qd8_f32_qc4w %[[LHS]], %[[RHS_UI4]] transpose_rhs = false : (tensor<100x200xi8>, tensor<300x200xi4>) -> tensor<100x300xf32>
+func.func @fully_connected(%input : tensor<100x200xi8>, %kernel : tensor<300x200xi4>) -> tensor<100x300xf32> {
   %kernel_cast = stablehlo.convert %kernel : (tensor<300x200xi4>) -> tensor<300x200xi8>
-  %dot_general = stablehlo.dot_general %input, %kernel_cast, contracting_dims = [2] x [1], precision = [DEFAULT, DEFAULT] : (tensor<1x100x200xi8>, tensor<300x200xi8>) -> tensor<1x100x300xi32>
-  %out = stablehlo.convert %dot_general : (tensor<1x100x300xi32>) -> tensor<1x100x300xf32>
-  return %out : tensor<1x100x300xf32>
+  %dot_general = stablehlo.dot_general %input, %kernel_cast, contracting_dims = [1] x [1], precision = [DEFAULT, DEFAULT] : (tensor<100x200xi8>, tensor<300x200xi8>) -> tensor<100x300xi32>
+  %out = stablehlo.convert %dot_general : (tensor<100x300xi32>) -> tensor<100x300xf32>
+  return %out : tensor<100x300xf32>
 }
 
 // CHECK-LABEL:   func.func @fully_connected$transpose(
-// CHECK-SAME:                                     %[[LHS:.*]]: tensor<1x100x200xi8>,
-// CHECK-SAME:                                     %[[RHS:.*]]: tensor<200x300xi4>) -> tensor<1x100x300xf32> {
+// CHECK-SAME:                                     %[[LHS:.*]]: tensor<100x200xi8>,
+// CHECK-SAME:                                     %[[RHS:.*]]: tensor<200x300xi4>) -> tensor<100x300xf32> {
 // CHECK:           %[[C8:.*]] = stablehlo.constant dense<8>
 // CHECK:           %[[C8_I4:.*]] = stablehlo.convert %[[C8]] : (tensor<{{.*}}xi8>) -> tensor<{{.*}}xi4>
 // CHECK:           %[[RHS_UI4:.*]] = stablehlo.xor %[[RHS]], %[[C8_I4]]
-// CHECK:           %{{.*}} = xnnpack.fully_connected_nc_qd8_f32_qc4w %[[LHS]], %[[RHS_UI4]] transpose_rhs = true : (tensor<1x100x200xi8>, tensor<200x300xi4>) -> tensor<1x100x300xf32>
-func.func @fully_connected$transpose(%input : tensor<1x100x200xi8>, %kernel : tensor<200x300xi4>) -> tensor<1x100x300xf32> {
+// CHECK:           %{{.*}} = xnnpack.fully_connected_nc_qd8_f32_qc4w %[[LHS]], %[[RHS_UI4]] transpose_rhs = true : (tensor<100x200xi8>, tensor<200x300xi4>) -> tensor<100x300xf32>
+func.func @fully_connected$transpose(%input : tensor<100x200xi8>, %kernel : tensor<200x300xi4>) -> tensor<100x300xf32> {
   %kernel_cast = stablehlo.convert %kernel : (tensor<200x300xi4>) -> tensor<200x300xi8>
-  %dot_general = stablehlo.dot_general %input, %kernel_cast, contracting_dims = [2] x [0], precision = [DEFAULT, DEFAULT] : (tensor<1x100x200xi8>, tensor<200x300xi8>) -> tensor<1x100x300xi32>
+  %dot_general = stablehlo.dot_general %input, %kernel_cast, contracting_dims = [1] x [0], precision = [DEFAULT, DEFAULT] : (tensor<100x200xi8>, tensor<200x300xi8>) -> tensor<100x300xi32>
+  %out = stablehlo.convert %dot_general : (tensor<100x300xi32>) -> tensor<100x300xf32>
+  return %out : tensor<100x300xf32>
+}
+
+// CHECK-LABEL:   func.func @fully_connected$rank3_input(
```

Comment:
Do we have a non-1 batch dimension case, like 8? Is it a case that can be supported? (I don't think we will have such a case since the xnnpack op says 'nc', and 'n' means the batch dimension, but I am curious if we have a batched fully-connected case.)

Reply:
> Is it a case that can be supported? 

Not with the fully connected op. AFAIK there is no 3d input case. There is the batch matrix multiplication op in XNNPACK, but this op does not have an Operator API.

None of the dispatches we are interested in at the moment require 3D inputs, but if it's needed, we can just ask the XNNPACK team for the Operator API.